src/stream/src/executor/source/source_executor.rs

```diff
@@ -610,6 +650,59 @@ impl<S: StateStore> Debug for SourceExecutor<S> {
     }
 }
 
+struct WaitEpochWorker<S: StateStore> {
+    wait_epoch_rx: UnboundedReceiver<(Epoch, HashMap<SplitId, String>)>,
+    state_store: S,
+}
+
+impl<S: StateStore> WaitEpochWorker<S> {
+    pub async fn run(mut self) {
+        tracing::debug!("wait epoch worker start success");
+        loop {
+            // poll the rx and wait for the epoch commit
+            match self.wait_epoch_rx.recv().await {
+                Some((epoch, updated_offsets)) => {
+                    tracing::debug!("start to wait epoch {}", epoch.0);
+                    let ret = self
+                        .state_store
+                        .try_wait_epoch(HummockReadEpoch::Committed(epoch.0))
+                        .await;
+
+                    match ret {
+                        Ok(()) => {
+                            tracing::debug!(epoch = epoch.0, "wait epoch success");
+                            // cdc source only has one split
+                            assert_eq!(1, updated_offsets.len());
+                            let (split_id, offset) = updated_offsets.into_iter().next().unwrap();
+                            let source_id: u64 = u64::from_str(split_id.as_ref()).unwrap();
+                            // notify cdc connector to commit offset
+                            match jni_source::commit_cdc_offset(source_id, offset.clone()) {
+                                Ok(_) => {}
+                                Err(e) => {
+                                    tracing::error!(
+                                        error = %e.as_report(),
+                                        "source#{source_id}: failed to commit cdc offset: {offset}.",
+                                    )
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            tracing::error!(
```

Comment:
At the first glance I thought we may miss `commit_cdc_offset` if we reach here. But I realize that `try_wait_epoch` on committed epoch won't fail unless the node is exiting. We may want to add a commment here for this assumption.

Reply:
I agree that as long as we don't miss offset commit for a long time, it is not a big deal. However, my point is not about whether we will miss it. My point is that `try_wait_epoch` is not expected to fail in this case. I originally thought we should panic here but logging an error log is more robust. How about just tweaking the message to be "Unexpected failure in try_wait_epoch {}" and adding a comment for this assumption?